src/ui/hooks/FlowRouter.tsx

```diff
@@ -0,0 +1,205 @@
+    const flow = useRef<Flow>(createFlow(0));
+    const componentStore = useRef<ComponentStore>(initComponentStore());
+
+    const [state, dispatch] = useReducer(flowReducer, {
+        step: 'confirmation',
+        authIndex: 0
+    });
+
+    const prevStep: string = flow.current[state.step]![FlowDispatchTypes.BACK] ?? '';
+
+    let TheComponent = renderScreen(state.step),
+        componentStoreMethods: ComponentStoreMethods = createComponentStoreConnection(state.step);
+
+    useEffect(() => {
+        flow.current = createFlow(state.authIndex);
+    }, [state.authIndex]);
+
+    useEffect(() => {
+        // eslint-disable-next-line
+        componentStoreMethods = createComponentStoreConnection(state.step);
+    }, [state.step, createComponentStoreConnection]);
+
+    function flowReducer(state: any, action: FlowAction): any {
+        const {type, payload} = action;
+        const {step} = state;
+        const theFlow: Flow = flow.current;
+
+        if (!theFlow.hasOwnProperty(step)) throw new Error(`Referenced step '${step}' does not exist in the flow`);
+
+        switch (type) {
+        case FlowDispatchTypes.NEXT:
+            return {
+                ...state,
+                step: theFlow[step]![FlowDispatchTypes.NEXT]
+            };
+        case FlowDispatchTypes.BACK:
+            return {
+                ...state,
+                step: theFlow[step]![FlowDispatchTypes.BACK]
+            };
+        case FlowDispatchTypes.RESTART:
+            return {
+                ...state,
+                step: 'confirmation'
+            };
+        case FlowDispatchTypes.SET_AUTH_METHOD:
+            if ('menu' === state.step) {
+                return {
+                    ...state,
+                    authIndex: payload
+                };
+            } else {
+                throw new Error('Please don\'t try to set the authentication method from outside the Authentication Menu screen');
+            }
+        default:
+            throw new Error('Unknown action type');
+        }
+    }
+
+    function renderScreen(step: string) {
+        const componentMap: ComponentMap = {
+            confirmation: 'ConfirmationScreen',
+            menu: 'AuthenticationOptionMenu',
+            details: 'ResultDetails',
+            agency_qr: 'AgencyQR',
+            email_input: 'EmailScreen',
+            smsotp: 'SMSOTPScreen',
+            verificationRequirement: 'VerificationRequirementScreen'
+        };
+
+        const component: any = React.lazy(() => import('../screens/' + componentMap[step]));
+
+        return component;
+    }
+
+    // TODO: Make into a hook and inject the component store as a parameter
+    function createComponentStoreConnection(step: string) {
+        const get: ComponentStoreGet = (dataKey: string, dfault?: any, component?: string) => {
+            // This is equivalent to component ??= step, but that broke the compiler
+            component ?? (component = step);
```

Comment:
I'm not sure I agree that this should be extendable. Personal opinion is that the `get` method should be fairly restrictive i.e. adding additional parameters should be painful.

I'm also thinking that
```
if (arguments[2] === null || arguments[2] === undefined {
  dfault = undefined;
}

if (arguments[2] === null || arguments[2] === undefined {
  component = step;
}
```

is not terribly more readable and much more verbose

Reply:
Now that I'm thinking about it, though, I'm not sure that `nullish` is even the right operator for the `component` parameter. The string shouldn't ever be empty, so it's probably better to write it as:

```
!component && (component = step);
```

Reply:
If the objective is to make the code harder to change by manually obfuscating the source code we will have to answer that we can't outcompete an obfuscater tool, and since we are open source we need to defend that contradiction with the future collaborators.

Reply:
My challenge to the line and your comment above is that '??' here is not used as intended and it forces you to wrap the right operand in (). If we accept verbose as only having more characters, this is better. But 'const key = component ?? step'